File: Assets/Scripts/CSharp2Lua/LuaCallEvent.cs
using UnityEngine;

public delegate void EventLua();

public class TestEvent
{
    public static event EventLua StaticEvent;
    public static void StaticFunc()
    {
        Debug.Log("这是C#静态事件触发");
    }
    public static void CallStatic()
    {
        if (StaticEvent != null)
            StaticEvent();
    }
    
    public event EventLua DelegateEvent;
    public void CallDynamic()
    {
        if (DelegateEvent != null)
            DelegateEvent();
    }
}

public class LuaCallEvent : MonoBehaviour
{
    private void Start()
    {
        XLuaEnv.Instance.DoString("CSharp2Lua/LuaCallEvent");
    }

    private void OnDestroy()
    {
        XLuaEnv.Instance.Dispose();
    }
}

File: DataPath/Lua/CSharp2Lua/LuaCallEvent.lua
-- Lua调用事件

-- C#添加事件 TestEvent.StaticEvent += TestEvent.StaticFunc



-- Lua添加事件
CS.TestEvent.StaticEvent("+", CS.TestEvent.StaticFunc);
CS.TestEvent.CallStatic();
CS.TestEvent.StaticEvent("-", CS.TestEvent.StaticFunc);



-- 添加动态成员变量
local func = function ()
    print("这是Lua调用C#动态事件触发");
end

local obj = CS.TestEvent();
obj:DelegateEvent("+", func);
obj:CallDynamic();
obj:DelegateEvent("+", func);